### PR TITLE
[Merged by Bors] - feat(RingTheory): define two-sided Jacobson radical

### PR DESCRIPTION
**Pull request by Vtec234**
- Add a constructor `TwoSidedIdeal.ofIdeal` to turn an `Ideal` that happens to be two-sided into a `TwoSidedIdeal`.
- Use `ofIdeal` to define a two-sided version of the Jacobson radical.
- Generalize some ring lemmas that didn't need to assume commutativity.
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/RingTheory/TwoSidedIdeal/Operations.lean [hunk 1]
```diff
@@ -248,6 +244,31 @@ lemma mem_asIdealOpposite {I : TwoSidedIdeal R} {x : Rᵐᵒᵖ} :
...
+lemma coe_ofIdeal (I : Ideal R) (h) : (ofIdeal I h : Set R) = I := by
+  simp [ofIdeal]
+
+@[simp]
+lemma ofIdeal_asIdeal (I : TwoSidedIdeal R) (h) : ofIdeal (asIdeal I) h = I := by
+  ext
+  simp [mem_ofIdeal, mem_asIdeal]
+
+@[simp]
+lemma asIdeal_ofIdeal (I : Ideal R) (h) : asIdeal (ofIdeal I h) = I := by
```
Comment (Vtec234) [inline Mathlib/RingTheory/TwoSidedIdeal/Operations.lean:265]: These could potentially be packaged as `TwoSidedIdeal R ≃o { I : Ideal R // mul_mem_right }`. However, I am not sure when (if ever) one would want to use that subtype instead of just `TwoSidedIdeal`.

File: Mathlib/RingTheory/TwoSidedIdeal/Basic.lean [hunk 2]
```diff
@@ -156,17 +156,16 @@ def mk' (carrier : Set R)
...
 
-lemma mem_mk' (carrier : Set R)
-    (zero_mem : 0 ∈ carrier)
-    (add_mem : ∀ {x y}, x ∈ carrier → y ∈ carrier → x + y ∈ carrier)
-    (neg_mem : ∀ {x}, x ∈ carrier → -x ∈ carrier)
-    (mul_mem_left : ∀ {x y}, y ∈ carrier → x * y ∈ carrier)
-    (mul_mem_right : ∀ {x y}, x ∈ carrier → x * y ∈ carrier)
-    (x : R) :
-    x ∈ mk' carrier zero_mem add_mem neg_mem mul_mem_left mul_mem_right ↔ x ∈ carrier := by
+lemma mem_mk' (carrier : Set R) (h₁ h₂ h₃ h₄ h₅) (x : R) :
```
Comment (Vtec234) [inline Mathlib/RingTheory/TwoSidedIdeal/Basic.lean:159]: Should the `mem_blah` lemmas be `@[simp]`?

File: Mathlib/RingTheory/JacobsonIdeal.lean [hunk 3]
```diff
@@ -214,6 +225,45 @@ theorem jacobson_mono {I J : Ideal R} : I ≤ J → I.jacobson ≤ J.jacobson :=
...
+    ∀ {x y}, x ∈ I.jacobson → x * y ∈ I.jacobson := by
+  -- Proof generalized from
+  -- https://ysharifi.wordpress.com/2022/08/16/the-jacobson-radical-definition-and-basic-results/
+  intro x r xJ
+  apply mem_sInf.mpr
+  intro 𝔪 𝔪_mem
+  by_cases r𝔪 : r ∈ 𝔪
+  · apply 𝔪.smul_mem _ r𝔪
+  let 𝔪₀ : Ideal R := {
+    carrier := { a : R | a*r ∈ 𝔪 }
```
Comment (Vtec234) [inline Mathlib/RingTheory/JacobsonIdeal.lean:239]: Is there a mathlib name for this construction?

Comment (eric-wieser) [reply Mathlib/RingTheory/JacobsonIdeal.lean:239]: This looks like it's the `Submodule.comap` by `DistribMulAction.toLinearMap` applied to `MulOpposite R`.

File: Mathlib/RingTheory/TwoSidedIdeal/Basic.lean (hunk 2 above)
Comment (eric-wieser) [inline Mathlib/RingTheory/TwoSidedIdeal/Basic.lean:159]: Why the change to the variable names?

Comment (Vtec234) [reply Mathlib/RingTheory/TwoSidedIdeal/Basic.lean:159]: The change to names is inconsequential (though I do think it's more readable not to have so many letters in the statement), but the change to their _types_ (namely, omitting the types and letting Lean infer them) is what let me simplify the proof of `mem_ker`. The difference in types is just whether something is eta-expanded, but the new version allows Lean to infer them and the previous one didn't.

File: Mathlib/RingTheory/JacobsonIdeal.lean (hunk 3 above)
Comment (Vtec234) [reply Mathlib/RingTheory/JacobsonIdeal.lean]: Thanks! Indeed this worked. I'm not sure it's a change for the better, but it does make the proof shorter. I left a comment explaining what the set actually is.

File: Mathlib/RingTheory/TwoSidedIdeal/Basic.lean (hunk 2 above)
Comment (jcommelin) [reply Mathlib/RingTheory/TwoSidedIdeal/Basic.lean:159]: This seems like a good one to be simp

File: Mathlib/RingTheory/TwoSidedIdeal/Operations.lean [hunk 4]
```diff
@@ -248,6 +244,31 @@ lemma mem_asIdealOpposite {I : TwoSidedIdeal R} {x : Rᵐᵒᵖ} :
   simpa [asIdealOpposite, asIdeal, TwoSidedIdeal.mem_iff, RingCon.op_iff] using
     ⟨I.ringCon.symm, I.ringCon.symm⟩
 
+/-- Bundle an `Ideal` that is already two-sided as a `TwoSidedIdeal`. -/
+def ofIdeal (I : Ideal R) (mul_mem_right : ∀ {x y}, x ∈ I → x * y ∈ I) : TwoSidedIdeal R :=
+  TwoSidedIdeal.mk' I I.zero_mem I.add_mem I.neg_mem (I.smul_mem _) mul_mem_right
+
+lemma mem_ofIdeal {I : Ideal R} {h} {x : R} :
```
Comment (jcommelin) [inline Mathlib/RingTheory/TwoSidedIdeal/Operations.lean:251]: And this one as well.

File: Mathlib/RingTheory/TwoSidedIdeal/Basic.lean (hunk 2 above)
Comment (eric-wieser) [reply Mathlib/RingTheory/TwoSidedIdeal/Basic.lean:160]: I guess the eta expanding could be avoided by writing `@one_mem` etc instead, but your version is certainly easier to read. The fact that eta-expansions are performed (presumably thanks to implicit lambdas) is quite surprising to me.

File: Mathlib/RingTheory/JacobsonIdeal.lean [hunk 5]
```diff
@@ -375,3 +421,16 @@ theorem isPrimary_of_isMaximal_radical [CommRing R] {I : Ideal R} (hi : IsMaxima
...
+
+namespace TwoSidedIdeal
+
+variable {R : Type u} [Ring R]
+
+/-- The Jacobson radical of `I` is the infimum of all maximal (left) ideals containing `I`. -/
+def jacobson (I : TwoSidedIdeal R) : TwoSidedIdeal R :=
+  TwoSidedIdeal.ofIdeal (asIdeal I).jacobson (Ideal.jacobson_mul_mem_right <| I.mul_mem_right _ _)
+
+lemma jacobson_asIdeal (I : TwoSidedIdeal R) : asIdeal I.jacobson = (asIdeal I).jacobson := by
```
Comment (j-loreaux) [inline Mathlib/RingTheory/JacobsonIdeal.lean:433]: or switch the order of the equality.
```suggestion
lemma asIdeal_jacobson (I : TwoSidedIdeal R) : asIdeal I.jacobson = (asIdeal I).jacobson := by
```

File: Mathlib/RingTheory/TwoSidedIdeal/Basic.lean (hunk 2 above)
Comment (j-loreaux) [reply Mathlib/RingTheory/TwoSidedIdeal/Basic.lean:160]: I would prefer to keep the names, please. Reasoning: `mk'` should actually be the default method for defining a `TwoSidedIdeal`, and users will probably need to apply `mem_mk'` soon thereafter. It would be a shame if they had to read the actual types of the hypotheses in order to do so, as opposed to just seeing the names and knowing intuitively what they are.

File: Mathlib/RingTheory/TwoSidedIdeal/Operations.lean [hunk 6]
```diff
@@ -248,6 +245,32 @@ lemma mem_asIdealOpposite {I : TwoSidedIdeal R} {x : Rᵐᵒᵖ} :
   simpa [asIdealOpposite, asIdeal, TwoSidedIdeal.mem_iff, RingCon.op_iff] using
     ⟨I.ringCon.symm, I.ringCon.symm⟩
 
+/-- Bundle an `Ideal` that is already two-sided as a `TwoSidedIdeal`. -/
+def ofIdeal (I : Ideal R) (mul_mem_right : ∀ {x y}, x ∈ I → x * y ∈ I) : TwoSidedIdeal R :=
+  TwoSidedIdeal.mk' I I.zero_mem I.add_mem I.neg_mem (I.smul_mem _) mul_mem_right
```
Comment (j-loreaux) [inline Mathlib/RingTheory/TwoSidedIdeal/Operations.lean:250]: I think that this name is not meaningfully different from `fromIdeal` (i.e., without checking the types, I won't know which one is which). I think `_root_.Ideal.toTwoSided` or `_root_.Ideal.toTwoSidedIdeal` would be better. I realize that this name might also suffer slightly from the problem I just mentioned *but*, unlike `fromIdeal`, which can't take advantage of dot notation because it's an order hom, this declaration can do so. This could be the distinguishing feature.

Obviously lemmas below would need to be renamed too, and I'm open to other suggestions here.

File: Mathlib/RingTheory/JacobsonIdeal.lean [hunk 7]
```diff
@@ -214,6 +225,41 @@ theorem jacobson_mono {I J : Ideal R} : I ≤ J → I.jacobson ≤ J.jacobson :=
   erw [mem_sInf] at hx ⊢
   exact fun K ⟨hK, hK_max⟩ => hx ⟨Trans.trans h hK, hK_max⟩
 
+theorem jacobson_mul_mem_right {I : Ideal R}
+    (mul_mem_right : ∀ {x y}, x ∈ I → x * y ∈ I) :
```
Comment (j-loreaux) [inline Mathlib/RingTheory/JacobsonIdeal.lean:229]: Please add a docstring pointing the user to `TwoSidedIdeal.jacobson` with a statement that they should prefer that when the ideal is two-sided.

File: Mathlib/RingTheory/TwoSidedIdeal/Basic.lean (hunk 2 above)
Comment (Vtec234) [reply Mathlib/RingTheory/TwoSidedIdeal/Basic.lean:160]: I readded the names, though I don't think anybody should be providing these arguments manually: they should be filled in by unification.

File: Mathlib/RingTheory/TwoSidedIdeal/Operations.lean (hunk 6 above)
Comment (Vtec234) [reply Mathlib/RingTheory/TwoSidedIdeal/Operations.lean:250]: I agree that having dot notation work is nice. For that reason, should `ofIdeal` also live in the `Ideal` namespace, and perhaps be renamed to `Ideal.twoSidedSpan` or something like that (in another PR!)?

Comment (j-loreaux) [reply Mathlib/RingTheory/TwoSidedIdeal/Operations.lean]: Maybe renaming it to `TwoSidedIdeal.spanIdeal` would make sense though.

**Review by j-loreaux — APPROVED** [review]
maintainer merge

**Review by jcommelin — COMMENTED** [review]
Thanks :tada:

bors merge

File: Mathlib/RingTheory/JacobsonIdeal.lean [hunk 8]
```diff
@@ -214,6 +225,44 @@ theorem jacobson_mono {I J : Ideal R} : I ≤ J → I.jacobson ≤ J.jacobson :=
   erw [mem_sInf] at hx ⊢
   exact fun K ⟨hK, hK_max⟩ => hx ⟨Trans.trans h hK, hK_max⟩
 
+/-- The Jacobson radical of a two-sided ideal is two-sided.
+
+It is preferable to use `TwoSidedIdeal.jacobson` instead of this lemma. -/
+theorem jacobson_mul_mem_right {I : Ideal R}
+    (mul_mem_right : ∀ {x y}, x ∈ I → x * y ∈ I) :
+    ∀ {x y}, x ∈ I.jacobson → x * y ∈ I.jacobson := by
```
Comment (alreadydone) [inline Mathlib/RingTheory/JacobsonIdeal.lean:233]: I found that [this unmerged mathlib3 PR](https://github.com/leanprover-community/mathlib3/pull/13862/files#diff-d50d7dee9b427f9710e98cc536adfd95f656e65f85f1ac179cd465c399375551R128-R129) actually contains this result (in the special case `I = 0`, but the proof probably generalizes, or else consider the quotient R/I).

Comment (Vtec234) [reply Mathlib/RingTheory/JacobsonIdeal.lean:233]: Do you think anything from that PR should be added to mathlib4, or are we happy with the API added here?

Comment (alreadydone) [reply Mathlib/RingTheory/JacobsonIdeal.lean:233]: I'll work on a proof that IsArtinianRing implies IsNoetherianRing starting tomorrow; we'll see what else I need.
